src/string-util/string-util.ts

```diff
@@ -15,50 +15,89 @@ const KOREA_LOT_NUMBER_ADDRESS_REGEXP = /읍|면|동|가/;
 const KOREA_ADDRESS_SORT_REGEXP = new RegExp(
   `${KOREA_ROAD_NAME_ADDRESS_REGEXP.source}|${KOREA_LOT_NUMBER_ADDRESS_REGEXP.source}`
 );
-const EMOJI_REGEXP =
-  /(\u00a9|\u00ae|[\u2000-\u3300]|\ud83c[\ud000-\udfff]|\ud83d[\ud000-\udfff]|\ud83e[\ud000-\udfff])/gi;
+
+//이모지 정규식 표현
+const EMOJI_REGEXP = /([\uD800-\uDBFF][\uDC00-\uDFFF])/;
+//ZERO WIDTH JOINER = 다중이모지에서 이모지 끼리 붙일 때 쓰이는 코드
+const EMOJI_ZERO_WIDTH_JOINER = '\u200D';
 
 export namespace StringUtil {
   export function maskPrivacy(text: string, type: PrivacyType): string {
-    const targetMask: Record<PrivacyType, MaskingRange> = {
-      name: {
-        start: EMOJI_REGEXP.test(text.slice(0, 2)) ? 2 : 1,
-        end: EMOJI_REGEXP.test(text.slice(text.length - 2)) ? text.length - 2 : text.length - 1,
-      },
-      phone: {
-        start: /\D/.test(text) ? getMaskingIndexOfUnnormalizedPhone(text).start : 3,
-        end: /\D/.test(text) ? getMaskingIndexOfUnnormalizedPhone(text).end : text.length - 4,
-      },
+    const maskingRule: Record<PrivacyType, MaskingRange> = {
+      name: { start: 1, end: text.length - 1 },
+      phone: { start: 3, end: text.length - 4 },
       email: { start: 2, end: text.indexOf('@') },
       bankAccount: { start: 3, end: text.length - 3 },
       address: { start: getMaskingStartIndexOfAddress(text), end: text.length },
     };
 
-    const start = targetMask[type].start;
-    const end = targetMask[type].end;
+    const start = maskingRule[type].start;
+    const end = maskingRule[type].end;
     const length = type === 'name' ? end - start || 1 : end - start;
-    return getMaskedString({ text, length, maskingStart: start });
+    return getMaskedString({ text, length, maskingStart: start, type });
+
+    function getMaskedString({ text, length, maskingStart, type }: MaskingOpts): string {
+      //emoji가 있는 text에 대해서 따로 처리
+      if (EMOJI_REGEXP.test(text)) {
+        return getMaskedWithEmoji(text);
+      }
 
-    function getMaskedString({ text, length, maskingStart }: MaskingOpts): string {
       const validLength = Math.max(0, length);
       const masking = '*'.repeat(validLength);
       const textSplit = [...text];
       textSplit.splice(maskingStart, validLength, masking);
+
       return textSplit.join('');
     }
 
-    function getMaskingIndexOfUnnormalizedPhone(phone: string): { start: number; end: number } {
-      const phoneNumberParts = phone.split(/\D/);
+    function getMaskedWithEmoji(text: string): string {
+      const splitNames = splitNameByEmoji(text);
+
+      let maskedName = '';
+
+      if (splitNames.length === 2) {
+        //이름이 2글자면 맨 뒤 마스킹
+        maskedName = splitNames[0] + '*';
+      } else {
+        //맨 앞과 맨 뒤 제외 마스킹
+        for (let i = 0; i < splitNames.length; i++) {
+          if (i === 0 || i === splitNames.length - 1) {
+            maskedName += splitNames[i];
+          } else {
+            maskedName += '*';
+          }
+        }
+      }
```

Comment:
(변경전 이름인)splitNameByEmoji의 결과를 생각했을 때 대충 이렇게 구현하면 될 거 같은데......
(테스트 안돌린 코드라 참고만 하세요)
```suggestion
      const charList = splitNameByEmoji(text);

      let maskedName = '';

      charList.forEach((char, index) => {
        if ((charList.length === 2 && index !== 0) || index !== charList.length - 1) {
          maskedName += '*';
        } else {
          maskedName += char;
        }
      });
```

Reply:
저도 Old Java에서 넘어오면서 적응이 잘 안되서 이해는 하지만
가급적 forEach, map 등의 array function들을 사용하는거에 익숙해지도록 하세요.
레거시 for문보다 이쪽이 성능이 보통 낫습니다.
(다만 쓸데없는 array function chaining은 지양해야합니다.)

Reply:
넵. 알려주셔서 감사합니다. 빨리 적응하겠습니다.